scylla-cdc/src/consumer.rs

```diff
@@ -0,0 +1,194 @@
+
+pub trait Consumer {
+    fn consume_cdc(&mut self, data: CDCRow);
+}
+
+pub trait ConsumerFactory {
+    fn new_consumer() -> Box<dyn Consumer>;
+}
+
+#[derive(Debug, Eq, PartialEq, TryFromPrimitive)]
+#[repr(i8)]
+pub enum OperationType {
+    PreImage,
+    RowUpdate,
+    RowInsert,
+    RowDelete,
+    PartitionDelete,
+    RowRangeDelInclLeft,
+    RowRangeDelExclLeft,
+    RowRangeDelInclRight,
+    RowRangeDelExclRight,
+    PostImage,
+}
+
+pub struct CDCRowSchema {
+    pub(crate) stream_id: usize,
+    pub(crate) time: usize,
+    pub(crate) batch_seq_no: usize,
+    pub(crate) end_of_batch: usize,
+    pub(crate) operation: usize,
+    pub(crate) ttl: usize,
+    pub(crate) mapping: HashMap<String, usize>,
+    pub(crate) deleted_mapping: HashMap<String, usize>,
+    pub(crate) deleted_el_mapping: HashMap<String, usize>,
+}
+
+const STREAM_ID_NAME: &str = "cdc$stream_id";
+const TIME_NAME: &str = "cdc$time";
+const BATCH_SEQ_NO_NAME: &str = "cdc$batch_seq_no";
+const END_OF_BATCH_NAME: &str = "cdc$end_of_batch";
+const OPERATION_NAME: &str = "cdc$operation";
+const TTL_NAME: &str = "cdc$ttl";
+const IS_DELETED_PREFIX: &str = "cdc$deleted_";
+const ARE_ELEMENTS_DELETED_PREFIX: &str = "cdc$deleted_elements_";
+
+impl CDCRowSchema {
+    pub fn new(specs: &[ColumnSpec]) -> CDCRowSchema {
+        // [TODO]:  This function is not usable now.
+        //          The driver does not provide the vector of ColumnSpec for every query yet,
+        //          and therefore we can't map the result
+        //          to something other than a hardcoded tuple.
+
+        let mut mapping: HashMap<String, usize> = HashMap::new();
+        let mut deleted_mapping: HashMap<String, usize> = HashMap::new();
+        let mut deleted_el_mapping: HashMap<String, usize> = HashMap::new();
+        let (mut stream_id, mut time, mut batch_seq_no, mut end_of_batch, mut operation, mut ttl):
+            (usize, usize, usize, usize, usize, usize) = (0, 0, 0, 0, 0, 0);
+        let mut j = 0;
+
+        // Hashmaps will have indices of data in a new vector without the hardcoded values.
+        for (i, spec) in specs.iter().enumerate() {
+            match spec.name.as_str() {
+                // spec.name is public since 0.3.0 driver version, it didn't work on 0.2.1.
+                STREAM_ID_NAME => stream_id = i,
+                TIME_NAME => time = i,
+                BATCH_SEQ_NO_NAME => batch_seq_no = i,
+                END_OF_BATCH_NAME => end_of_batch = i,
+                OPERATION_NAME => operation = i,
+                TTL_NAME => ttl = i,
+                x => {
+                    if x.starts_with(IS_DELETED_PREFIX) {
+                        deleted_mapping.insert(x.to_string(), j);
+                    } else if x.starts_with(ARE_ELEMENTS_DELETED_PREFIX) {
+                        deleted_el_mapping.insert(x.to_string(), j);
+                    } else {
+                        mapping.insert(x.to_string(), j);
+                    }
```

Comment:
By the way, you can write this logic without using any unwraps:

```rust
if let Some(stripped) = x.strip_prefix(ARE_ELEMENTS_DELETED_PREFIX) {
    deleted_el_mapping.insert(stripped.to_string(), j);
} else if let Some(stripped) = x.strip_prefix(IS_DELETED_PREFIX) {
    deleted_mapping.insert(stripped.to_string(), j);
} else {
    mapping.insert(x.to_string(), j);
}
```

Reply:
Additionally, please note that IS_DELETED_PREFIX itself is a prefix of ARE_ELEMENTS_DELETED_PREFIX - when  analyzing the variable name, you need to first check for the ARE_ELEMENTS_DELETED_PREFIX and only then for IS_DELETED_PREFIX.